File: back/app/models/connection.py
from django.db import models

from app.base.models.base_models import ActiveManager, BaseEntityModelMixin
from app.models.comment import Comment
from app.models.project import Project
from app.util import format_decimal
from documents.base.base_models import BaseDocumentModel


# TO-DO: Harcoded nº of people per household
NUMBER_OF_PEOPLE_PER_HOUSEHOLD = 5


class Connection(BaseDocumentModel, BaseEntityModelMixin):
    class Meta(object):
        db_table = "connection"
        verbose_name = "Conexión"
        verbose_name_plural = "Conexiones"

    objects = ActiveManager()

    number_of_households = models.IntegerField("Número de viviendas", null=True)
    number_of_existing_connections = models.IntegerField(
        "Número de conexiones existentes", null=True
    )
    number_of_planned_connections = models.IntegerField(
        "Número de conexiones previstas", null=True
    )
    number_of_actual_connections = models.IntegerField(
        "Número de conexiones reales", null=True
    )
    project = models.ForeignKey(
        Project,
        on_delete=models.CASCADE,
        verbose_name=Project._meta.verbose_name,
        related_name="connections",
    )
    comments = models.ManyToManyField(Comment)

    @property
    def population(self):
        if not self.number_of_households:
            return None
        return self.number_of_households * NUMBER_OF_PEOPLE_PER_HOUSEHOLD

    @property
    def connected_households_percentage(self):
        number_of_households = self.number_of_households
        number_of_actual_connections = self.number_of_actual_connections

        if not number_of_households or not number_of_actual_connections:
            return 0
        return format_decimal(
            (number_of_actual_connections / number_of_households) * 100
        )

    @property
    def coverage(self):
        if not self.number_of_actual_connections:
            return None
        return self.number_of_actual_connections * NUMBER_OF_PEOPLE_PER_HOUSEHOLD

    def __str__(self):
        return self.id


def create_project_connection(project):
    connection = Connection(
        project=project,
        created_by=project.creation_user,
        updated_by=project.creation_user,
    )
    connection.save()

File: front/src/connection/model/Connection.js
import {comments_api_adapter, createComments} from "comment/model";

class Connections extends Array {}

const connection_api_adapter = connection => {
    if (connection["comments"]) {
        connection["comments"] = createComments(
            comments_api_adapter(connection["comments"])
        );
    }

    return connection;
};

const connection_view_adapter = connection => {
    delete connection["folder"];
    delete connection["created_by"];
    delete connection["created_at"];
    delete connection["updated_at"];
    delete connection["updated_by"];

    delete connection["population"];
    delete connection["connected_households_percentage"];
    delete connection["coverage"];
    delete connection["connection_comments"];

    return connection;
};

const connections_api_adapter = connections => {
    return connections.map(connection_api_adapter);
};

const createConnections = (data = []) => {
    const connections = Connections.from(data, connection =>
        createConnection(connection)
    );
    return connections;
};

const createConnection = ({
    id = null,
    number_of_households = null,
    population = null,
    number_of_existing_connections = null,
    number_of_planned_connections = null,
    number_of_actual_connections = null,
    connected_households_percentage = null,
    coverage = null,
    comments = [],
    project = null,
    folder = "",
    created_by = "",
    created_at = null,
    updated_at = null,
    updated_by = "",
} = {}) => {
    const publicApi = {
        id,
        number_of_households,
        population,
        number_of_existing_connections,
        number_of_planned_connections,
        number_of_actual_connections,
        connected_households_percentage,
        coverage,
        project,
        comments,
        folder,
        created_by,
        created_at,
        updated_at,
        updated_by,
    };

    return Object.freeze(publicApi);
};

export {
    createConnection as default,
    createConnections,
    connection_api_adapter,
    connections_api_adapter,
    connection_view_adapter,
};
